nanoc/lib/nanoc/base/repos/binary_compiled_content_cache.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module Nanoc
+  module Int
+    # Represents a cache than can be used to store already compiled content,
+    # to prevent it from being needlessly recompiled.
+    #
+    # @api private
+    class BinaryCompiledContentCache < ::Nanoc::Int::Store
+      include Nanoc::Core::ContractsSupport
+
+      contract C::KeywordArgs[config: Nanoc::Core::Configuration] => C::Any
+      def initialize(config:)
+        super(Nanoc::Int::Store.tmp_path_for(config: config, store_name: 'binary_content'), 1)
+
+        @cache = {}
+      end
+
+      contract Nanoc::Core::ItemRep => C::Maybe[C::HashOf[Symbol => Nanoc::Core::Content]]
+      # Returns the cached compiled content for the given item representation.
+      #
+      # This cached compiled content is a hash where the keys are the snapshot
+      # names. and the values the compiled content at the given snapshot.
+      def [](rep)
+        item_cache = @cache[rep.item.identifier] || {}
+
+        rep_cache = item_cache[rep.name]
+        return nil if rep_cache.nil?
+
+        rep_cache.transform_values do |filename|
+          Nanoc::Core::Content.create(filename, binary: true)
+        end
+      end
+
+      contract Nanoc::Core::ItemRep, C::HashOf[Symbol => Nanoc::Core::BinaryContent] => C::HashOf[Symbol => Nanoc::Core::Content]
+      # Sets the compiled content for the given representation.
+      #
+      # This cached compiled content is a hash where the keys are the snapshot
+      # names. and the values the compiled content at the given snapshot.
+      def []=(rep, content)
+        @cache[rep.item.identifier] ||= {}
+        @cache[rep.item.identifier][rep.name] ||= {}
+        rep_cache = @cache[rep.item.identifier][rep.name]
+
+        content.each do |snapshot, binary_content|
+          filename = build_filename(rep, snapshot)
+          rep_cache[snapshot] = filename
+
+          # ???
+          next if binary_content.filename == filename
+
+          # TODO: hardlink, if possible
```

Comment:
This is wrong — hardlinks will not behave properly as it can lead to incorrect situations where one of the hardlinked files change.